apollo-federation/src/merger/merge_field.rs

```diff
@@ -696,6 +726,115 @@ impl Merger {
+
+        fn print_subgraphs<T: HasSubgraph>(arr: &[T]) -> String {
+            human_readable_subgraph_names(arr.iter().map(|s| s.subgraph()))
+        }
+
+        if !non_shareable_sources.is_empty()
+            && (!shareable_sources.is_empty() || non_shareable_sources.len() > 1)
+        {
+            let resolving_subgraphs = print_subgraphs(&all_resolving);
+            let non_shareables = if shareable_sources.is_empty() {
+                "all of them".to_string()
+            } else {
+                print_subgraphs(&non_shareable_sources)
+            };
+
+            // An easy-to-make error that can lead here is the misspelling of the `from` argument of an @override. Because in that case, the
+            // @override will essentially be ignored (we'll have logged a warning, but the error we're about to log will overshadow it) and
+            // the 2 field instances will violate the sharing rules. But because in that case the error is ultimately with @override, it
+            // can be hard for user to understand why they get a shareability error, so we detect this case and offer an additional hint
+            // at what the problem might be in the error message (note that even if we do find an @override with a unknown target, we
+            // cannot be 100% sure this is the issue, because this could also be targeting a subgraph that has just been removed, in which
+            // case the shareable error is legit; so keep the shareability error with a strong hint is hopefully good enough in practice).
+            // Note: if there is multiple non-shareable fields with "target-less overrides", we only hint about one of them, because that's
+            // easier and almost surely good enough to bring the attention of the user to potential typo in @override usage.
+            let subgraph_with_targetless_override = non_shareable_sources
+                .iter()
+                .find(|s| merge_context.has_override_with_unknown_target(s.idx));
+
+            let extra_hint = if let Some(s) = subgraph_with_targetless_override {
+                format!(
+                    " (please note that \"{}.{}\" has an @override directive in {} that targets an unknown subgraph so this could be due to misspelling the @override(from:) argument)",
+                    dest.type_name(),
+                    dest.field_name(),
+                    s.subgraph,
+                )
+            } else {
+                "".to_string()
+            };
+            self.error_reporter.add_error(CompositionError::InvalidFieldSharing {
+                message: format!(
+                    "Non-shareable field \"{}.{}\" is resolved from multiple subgraphs: it is resolved from {} and defined as non-shareable in {}{}",
+                    dest.type_name(),
+                    dest.field_name(),
+                    resolving_subgraphs,
+                    non_shareables,
+                    extra_hint,
+                ), // TODO: Add nodes
```

Comment:
You probably need to rebase on top of dev. There is a new lint that will inline the variables.